EnrichmentScope & EnrichingActivityProcessor

CodeBlanch opened the pull request:
Relates to #73. Relates to #809 (option 3, `StartAction`).

## Changes

Added the concept of an enrichment scope. The idea is to give users the ability to add contextual data to the Activity created performing some operation, in a global way.

* [X] Design discussion (this).
* [X] TODOs.
* [X] Unit tests.
* [X] CHANGELOG update.
* [X] Example + README.

CodeBlanch commented on examples/AspNetCore/Controllers/WeatherForecastController.cs:
```diff
@@ -19,8 +20,14 @@ public class WeatherForecastController : ControllerBase
-        public IEnumerable<WeatherForecast> Get()
+        public IEnumerable<WeatherForecast> Get(int? zipPostalCode = null)
         {
+            using var scope = EnrichmentScope.Begin(a =>
```
This demos usage. Here the user wants to add "user.zip_postal" tag, if provided, on the HTTP span that could be created (enrichment only happens if the activity is created & sampled).

cijothomas replied:
Why don't we let users do Activity.Current?.AddTag(AddTag("user.zip_postal", zipPostalCode.Value.ToString())

CodeBlanch replied:
@cijothomas That's what this is doing 😄

* If the user does `Activity.Current?.AddTag("user.zip_postal", zipPostalCode.Value.ToString())` the data shows up on the currently running span. In the case of this code, that will be the http-in span.

* If the user does `EnrichmentScope.Begin` the data shows up on the NEXT span, aka child span, which does not exist yet when the code is executing. `EnrichmentScope` is basically allowing you to hang on to that contextual information until the Activity is available. In the case of this code, that will be the http-out span.

alanwest replied:
I do like the concept. I've had customers want to query their data before and be able to group it by a combination of attributes on an automatically created span (like the http-out span) and attributes from one of their custom spans, though depending on what your backend is capable of this may not be easily possible if the attributes are spread across two spans.

This does mean that _all_ the spans created within the scope get "enriched" right? So, for instance in the case of gRPC which uses HTTP under the covers, both the gRPC span and HTTP span would get the `user.zip_postal` attribute. Well.. then again, I suppose the `EnrichmentAction` could be arbitrarily complex and you could make sure you target the correct span.  Maybe that's ok, but it may get a bit complicated since writing the `EnrichmentAction` depends on the user knowing what the instrumentation is doing.

All that said, yes I do like this, and while sure it could sometimes become complicated or lead to unintuitive results, I think there are many use cases this would benefit greatly.

CodeBlanch replied:
> This does mean that _all_ the spans created within the scope get "enriched" right?

No! The next Activity/Span created+sampled is enriched, then the scope is automatically closed. It does not flow to children. That is key. It's not meant to be baggage or correlation, there are already mechanisms/designs for that.

> As the recommended way to enrich current activity at any point in user code is to retrieve the current active span and enrich it, by doing tracer.CurrentSpan.SetAttribute or in .NET, Activity.Current?.SetAttribute.

Of course, agree 100%. But not always possible! EnrichmentScope is enabling a previously impossible scenario. Let me try to explain the user story for this.

I have some groups in my company working with OTel. I had the same ask from two different teams. Both expose an internal API that fronts a big database. Their spans are primarily HTTP-IN and SQL-OUT. Almost immediately both teams asked to enrich their SQL spans. The data (tags) added by the Sql Instrumentation are basic, taken off of the SqlConnection & SqlCommand. The code calling Sql has a bunch of _contextual_ information about what the library/API is doing. It knows the user, the area, the application, etc. That information is never passed on the SqlConnection or SqlCommand. The ask from these users was, how do we add this _contextual_ information on our SQL spans? Not possible. Both teams opted to turn off the Sql instrumentation and manually instrumented their code, adding their own data and re-doing what the instrumentation was doing.

The _contextual_ information is the key. You can't `Activity.Current.AddTag` that stuff because the Activity is created later, without the information around.

> The original issue #73 is asking for a way to enrich the current span with arbitrary information available from the httprequestmessage etc.

Yes. That could be done with options. You could get HttpRequest & HttpResponse, but not the contextual stuff. It also wouldn't help you with SQL. We could add SQL options, but you would just get SqlCommand & SqlConnection, which has limited utility.

What I'm doing here is a bit bigger than the ask on the issue. This solve is global, applies to any instrumentation the user might use, and solves the context thing, which was previously impossible.

But what if the user wanted the raw HttpRequest or raw HttpResponse? We could still have the options thing. I think all instrumentation should have basic settings for filters & enrichment. That doesn't undo the need for an enrichment scope thing though, because it's solving the _contextual_ information needing to be pushed on the next span.

By the way, using this enrichment pattern you could actually also solve the raw Http case. Like this:
```csharp
            using var scope = EnrichmentScope.Begin(a =>
            {
                if (zipPostalCode.HasValue)
                    a.AddTag("user.zip_postal", zipPostalCode.Value.ToString());

                HttpWebRequest request = (HttpWebRequest)a.GetCustomProperty("HttpWebRequest.Request");
                a.AddTag("http.content_type", request.ContentType);
            });
```

alanwest replied:
Nicely put. The use case you describe is exactly what I've come up against in the past (and did not describe nearly as well 😄). This arbitrary contextual information is beneficial with the customers I've encountered because New Relic provides a query interface for querying span data. With this it would be nice to be able to query `SELECT * FROM Span WHERE http.url LIKE "%blah.com%" AND zipCode = 90210`. This query isn't possible if the zip code attribute is not on the HTTP span.

Though I'm still thinking that this is a hard problem. It seems all too easy to make unintentional changes. Like what would happen if the code you have was refactored to:

```
            using var scope = EnrichmentScope.Begin(a =>
            {
                // The span generated downstream from this call would get decorated?
                NewMethodCallHereThatDoesSomeWorkWithADatabaseAndCreatesASpan();

                if (zipPostalCode.HasValue)
                    a.AddTag("user.zip_postal", zipPostalCode.Value.ToString());

                HttpWebRequest request = (HttpWebRequest)a.GetCustomProperty("HttpWebRequest.Request");
                a.AddTag("http.content_type", request.ContentType);
            });
```

CodeBlanch replied:
@alanwest Yes in that case the scope would end up wrong. Not something I am torn up about necessarily. I just feel like... _why_ would you make a call right there? Is it an edge case? We could actually (probably) support it. If we close the scope(s) before calling the callback(s).

> Another idea I've pondered in the past is the idea of a scope provided with a callback where the callback is invoked when spans are created allowing you to be smart about what it's enriching... haven't thought this through too deeply. The context you want to add may not be available in the scope of the callback...

Can you go into that a bit more? Kind of sounds like what I'm doing here.

alanwest replied:
> Not something I am torn up about necessarily. I just feel like... why would you make a call right there? Is it an edge case?

I'm not too concerned about it either, just raising it. Whether it is an edge case or not is hard to tell. Maybe a slightly more interesting example would be:

```csharp
            using var scope = EnrichmentScope.Begin(a =>
            {
                ...
            });

            // Maybe this method is owned by some other team or library out of my control.
            MethodThatMakesHttpOrSqlCallDownstreamThatIWantEnriched();
```

It's even possible that this code works one day, but then later the implementation changes and creates a span somewhere downstream.

I'm not suggesting we need to solve all the problems. I'm fine with limiting the situations we can feasibly support, I'm just thinking things through. I am 👍 on this idea and would be pretty interested in learning of any rough edges through people's experience.

> We could actually (probably) support it.

This does seem good if it can work.

---

> Can you go into that a bit more? Kind of sounds like what I'm doing here.

Ha! Sorry, I generated a jumbled 🕸️ of thought here. What was more on my mind was _not_ a scope, but maybe a delegate that HTTP instrumentation could be configured with that gets invoked on all HTTP calls enabling you to enrich at that point in time. But this idea definitely precludes you from accessing the important context you'd want to enrich with.

CodeBlanch replied:
> A delegate that HTTP instrumentation could be configured with that gets invoked on all HTTP calls enabling you to enrich at that point in time.

I'll do that too next. I'm all for adding points in the process users can hook into. IMO all the instrumentation should have standard delegates (more or less) for filters and enrichment. BUT, they are of limited utility. A delegate on Http options will get you HttpRequest, HttpResponse, and maybe Exception. There's only so much data on those that will be of interest. Something like our .NET Framework SqlClient, you only get a few strings. That's why I went for the enrichment scope, so users can add whatever they want from a closure. More power 🏭

alanwest replied:
A thought stemming from one of my earlier questions: "This does mean that all the spans created within the scope get "enriched" right?"

What if this were an option? Something like:

```csharp
enum EnrichmentScopeOptions
{
    FirstSpan, // default behavior - only enrich the next span
    AllSpans   // enrich all spans within scope
}
```

reyang replied:
or TTL 😄

alanwest replied:
Time to live?

CodeBlanch replied:
I just pushed new changes. Going to resolve this thread, please continue discussing though!

reyang commented on src/OpenTelemetry/Trace/EnrichmentScope.cs:
```diff
@@ -0,0 +1,78 @@
+    /// <summary>
+    /// A class for enriching the data of <see cref="Activity"/> objects.
+    /// </summary>
+    public class EnrichmentScope : IDisposable
```
I know I'm always bad at naming things, the name `ContextualProcessor` keeps popping up in my mind 😂

CodeBlanch replied:
(I'm not opposed to changing the name. I was just waiting to see if anyone else thought similar or had other ideas.)

reyang replied:
Should be my last question in this PR: do we envision a similar concept would be introduced for metrics/logs in the future? If yes, we might need to consider how we name things so we won't be cornered in the future.

alanwest replied:
This is an interesting point. Slight tangent, but my mental model of "instrumentation" before OTel has been one that combines the concepts of collecting of traces, metrics, etc together. High-level I think of something like HTTP instrumentation as a thing that both generates spans in a trace as well as collecting metrics about the operation. From a configuration perspective, I've wondered if these things will continue to be treated as separate - that is, I build up my trace pipeline/configuration separate from my metric configuration.

In the context of EnrichmentScope, should there be a different trace vs. metric vs. log EnrichmentScope concept? Or can all of these be combined in one EnrichmentScope class that includes the ability to configure behavior on traces, metrics, and logs?

CodeBlanch replied:
There's nothing really Activity-specific about the current implementation. It uses @reyang's `RuntimeContext` stuff. Essentially what it is doing is buffering a callback/closure to be executed when the final _thing_ to be enriched is available. So in that way, it could work for metrics & logs. Maybe you will be able to give it different callbacks depending on what is being enriched? I don't have enough of a mental picture of the metrics & logging APIs to say conclusively, but it would be great if we had one thing for all the spec areas. Open to suggestions.

reyang replied:
My suggestion is that we get this PR merged and continue the discussion here 😆

alanwest replied:
Yes, I'm leaning towards thinking one thing for all the spec areas make sense, though like you my mental picture is still somewhat fuzzy.

This is why I was pondering this thought in the realm of configuration since that's a little more concrete at the moment. For example, I can see the `AddOpenTelemetry` extension method on `IServiceCollection` include additional parameters for a "MeterProviderBuilder" and a "LogProviderBuilder" like how you're suggesting distinct callbacks based on type of telemetry data for EnrichmentScope.

alanwest replied:
> My suggestion is that we get this PR merged and continue the discussion here 😆

yes! agreed.

reyang commented on src/OpenTelemetry/Sdk.cs:
```diff
@@ -185,7 +185,28 @@ public static TracerProvider CreateTracerProvider(Action<TracerProviderBuilder>
-                ActivityStopped = activityProcessor.OnEnd,
+                ActivityStopped = (activity) =>
+                {
+                    EnrichmentScope scope = EnrichmentScope.Current;
```
I guess this can be put into an `EnrichmentActivityProcessor` which folks can optionally add/enable, instead of in the hot-path by default (that everyone has to pay the tax)?

CodeBlanch replied:
Great idea! I'll update soon.

CodeBlanch commented:
Updated the design:

* Added EnrichingActivityProcessor and moved the logic to apply EnrichmentScopes into there. This was a result of @reyang's feedback. Users not interested in the enrichment no longer need to pay for the perf.
* Added EnrichmentScopeTarget. This was a result of @alanwest's feedback about options for how the scope should be applied. A little tricky to implement, but I think I got it to work using a doubly-linked-list design.

Fire away!

reyang commented on src/OpenTelemetry/Trace/EnrichmentScope.cs:
```diff
@@ -0,0 +1,100 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
```
nit: add a blank line above this?

CodeBlanch replied:
Dang I even tried to be mindful of this 😄 Updated.

reyang replied:
I'll add another cop before we GA (and yet another cop to check CR/LF/CRLF and literal `TAB`s as the current repo is quite messy).

reyang commented on src/OpenTelemetry/Trace/EnrichmentScope.cs:
```diff
@@ -0,0 +1,100 @@
+            {
+                this.EnrichmentAction = null;
+
+                this.Parent.Child = this.Child;
```
For safety, do we want to check:
```csharp
if (RuntimeContextSlot.Get() != this)
{
    return; // do nothing
}
```

reyang replied:
No need to address in this PR, I have these in my TODO list:
* What would happen (worst case) if you start and end activities in different order.
* What would happen (worst case) if you end activity in a different thread.

CodeBlanch replied:
Check out what I just pushed. Tried to make it a bit more defensive and also have it "snap back" to the correct parent should things go out of order.

reyang replied:
LGTM.
If we move a step further, it'll be HAMT in [PEP0550](https://www.python.org/dev/peps/pep-0550/#implementation).

reyang commented on docs/trace/enriching-activity-processor/Program.cs:
```diff
@@ -0,0 +1,67 @@
+            using var request = new HttpRequestMessage
+            {
+                Method = HttpMethod.Get,
+                RequestUri = new Uri("https://www.google.com/"),
```
Would vote for "Wikipedia" to be "vendor agnostic" 🐇

reyang commented on docs/trace/enriching-activity-processor/Program.cs:
```diff
@@ -0,0 +1,67 @@
+        }
+
+        Console.WriteLine("Press any key to exit.");
+        Console.ReadKey();
```
Curious - why do we need this?

CodeBlanch replied:
When you launch via VS it closes instantly so you can't see the output.

reyang replied:
It feels a bit weird as we're inconsistent with other examples and also the ["Hello, World" tutorial](docs.microsoft.com/dotnet/core/tutorials/with-visual-studio#create-the-app).

reyang replied:
FYI - this [article](https://blog.bytellect.com/software-development/visual-studio/keeping-the-console-window-open-in-visual-studio-2017/) might help to solve the VS myth.

CodeBlanch replied:
Removed

reyang commented on docs/trace/enriching-activity-processor/README.md:
```diff
@@ -0,0 +1,113 @@
+
+Sometimes the built-in instrumentation doesn't add enough data or you want to
+augment the data it provides with contextual information your application or
+library has available. For these cases OpenTelemetry-dotnet provides an
```
```suggestion
library has available. For these cases OpenTelemetry .NET provides an
```

CodeBlanch commented:
@eddynaka @reyang Markdownlint is complaining about table rows being too long. How do I make that happy? Doesn't seem to be a way to break in the middle of a row.

reyang commented on docs/trace/enriching-activity-processor/README.md:
```diff
@@ -0,0 +1,113 @@
+library has available. For these cases OpenTelemetry-dotnet provides an
+`EnrichingActivityProcessor` and `EnrichmentScope` API.
+
+1. To enrich your spans first add the `EnrichingActivityProcessor` to your
```
I guess we want to have a consistent usage of span/activity across the docs/APIs/comments?

CodeBlanch replied:
It is a bit of a strange situation we've put ourselves in. Anyone using Activity API will want to see Activity. Anyone using OpenTelemetry API will want to see Span. I'll switch that one to Activity since the rest of the doc uses Activity terms, but I'm not sure which way it should really be.

reyang replied:
Yeah, it is a hard naming problem.
I thought about this while I was writing the other docs, given .NET has `Span` which is totally different concept, I think it gives us perfect excuse to avoid using `span` in our docs (except for the API conceptual doc where we elaborate the differences).

reyang commented on docs/trace/enriching-activity-processor/README.md:
````diff
@@ -0,0 +1,113 @@
+1. To enrich your spans first add the `EnrichingActivityProcessor` to your
+   `TracerProvider`:
+
+    ```csharp
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+                .AddHttpClientInstrumentation()
+                .AddProcessor(new EnrichingActivityProcessor())
+                .AddConsoleExporter()
+                .Build();
+    ```
````
```suggestion
    ```csharp
    using var tracerProvider = Sdk.CreateTracerProviderBuilder()
        .AddHttpClientInstrumentation()
        .AddProcessor(new EnrichingActivityProcessor())
        .AddConsoleExporter()
        .Build();
    ```
```

reyang commented on docs/trace/enriching-activity-processor/README.md:
````diff
@@ -0,0 +1,113 @@
+2. Use `EnrichmentScope.Begin` to wrap the call you want to instrument:
+
+    ```csharp
+            using (EnrichmentScope.Begin(a =>
````
Same here, remove the unnecessary indentation so it reads better on smaller devices/screens.

reyang commented on docs/trace/enriching-activity-processor/README.md:
````diff
@@ -0,0 +1,113 @@
+            }
+    ```
+
+    In that example the span created for the call to Google will be decorated
````
Consider "Wikipedia"?

eddynaka commented:
> @eddynaka @reyang Markdownlint is complaining about table rows being too long. How do I make that happy? Doesn't seem to be a way to break in the middle of a row.

don't know if you solved already. but you can try adding a <br>in the row to break and see if it complains.

reyang commented on docs/trace/enriching-activity-processor/README.md:
```diff
@@ -0,0 +1,113 @@
+    registered before your `Exporter` and any other `ActivityProcessor`s you are
+    using.
+
+2. Use `EnrichmentScope.Begin` to wrap the call you want to instrument:
```
For consideration: instead of using `2.`, use L2 heading (which helps to save the indentation and also gives anchor for people to link to.

reyang commented on docs/trace/enriching-activity-processor/README.md:
```diff
@@ -0,0 +1,113 @@
+
+In that example the request to `www.google.com` will be enriched with the
+`mycompany.user_id=1234` and `mycompany.customer_id=5678` tags (in that order)
+and the request to `www.bing.com` will be enriched with the
```
Last time I saw bing.com was 3 years ago 😸

reyang commented on docs/trace/enriching-activity-processor/README.md:
````diff
@@ -0,0 +1,113 @@
+    ```
+
+    **Note:** Order is important. Make sure your `EnrichingActivityProcessor` is
+    registered before your `Exporter` and any other `ActivityProcessor`s you are
````
```suggestion
    added before your `Exporter` and any other `ActivityProcessor`s you are
```

CodeBlanch commented:
@eddynaka 

> don't know if you solved already. but you can try adding a
in the row to break and see if it complains.

Add a what? 😄 I tried a few things, couldn't get it to work.

Here's the source:
```
| Name | Description |
| --- | --- |
| FirstChild | The first child `Activity` created under the scope will be enriched and then the scope will automatically be closed. |
```
Yields:
| Name | Description |
| --- | --- |
| FirstChild | The first child `Activity` created under the scope will be enriched and then the scope will automatically be closed. |

```
| Name | Description |
| --- | --- |
| FirstChild | The first child `Activity` created under the scope will
 be enriched and then the scope will automatically be closed. |
```
Yields:
| Name | Description |
| --- | --- |
| FirstChild | The first child `Activity` created under the scope will
 be enriched and then the scope will automatically be closed. |

```
| Name | Description |
| --- | --- |
| FirstChild | The first child `Activity` created under the scope will |
|            | be enriched and then the scope will automatically be closed. |
```
Yields:
| Name | Description |
| --- | --- |
| FirstChild | The first child `Activity` created under the scope will |
|                 | be enriched and then the scope will automatically be closed. |

I think we may need to alter the markdownlint config to ignore `tables`?

eddynaka commented:
> @eddynaka
> 
> > don't know if you solved already. but you can try adding a
> > in the row to break and see if it complains.
> 
> Add a what? 😄 I tried a few things, couldn't get it to work.
> 
> Here's the source:
> 
> ```
> | Name | Description |
> | --- | --- |
> | FirstChild | The first child `Activity` created under the scope will be enriched and then the scope will automatically be closed. |
> ```
> 
> Yields:
> 
> Name	Description
> FirstChild	The first child `Activity` created under the scope will be enriched and then the scope will automatically be closed.
> ```
> | Name | Description |
> | --- | --- |
> | FirstChild | The first child `Activity` created under the scope will
>  be enriched and then the scope will automatically be closed. |
> ```
> 
> Yields:
> 
> Name	Description
> FirstChild	The first child `Activity` created under the scope will
> be enriched and then the scope will automatically be closed.	
> ```
> | Name | Description |
> | --- | --- |
> | FirstChild | The first child `Activity` created under the scope will |
> |            | be enriched and then the scope will automatically be closed. |
> ```
> 
> Yields:
> 
> Name	Description
> FirstChild	The first child `Activity` created under the scope will
> be enriched and then the scope will automatically be closed.
> I think we may need to alter the markdownlint config to ignore `tables`?

I sent the tag in gitter...github replaced the tag with the new line LOL

reyang commented:
> @eddynaka @reyang Markdownlint is complaining about table rows being too long. How do I make that happy? Doesn't seem to be a way to break in the middle of a row.

Remove the table, use bullet points. In general markdownlint tries to discourage people from putting long statements in a table due to the readability issue.

alanwest commented on docs/trace/enriching-activity-processor/README.md:
````diff
@@ -0,0 +1,118 @@
+    .Build();
+```
+
+**Note:** Order is important. Make sure your `EnrichingActivityProcessor` is
+added before your `Exporter` and any other `ActivityProcessor`s you are using.
````
Not suggesting any action for this PR, but food for thought...

This note makes me wonder if we should consider codifying processor/exporter ordering in an effort to make this less prone to messing up. I don't have a clear suggestion at the moment, but maybe a combination of "type" and "priority".

Like if we know it's an "exporter type" versus "processor type", can the builder be smart enough to make sure to place it at the end?

Furthermore, maybe not all processors are equal. Some may be "high priority" and should therefore be injected at the beginning of the pipeline. This would in turn require some logic to deal with ties.

CodeBlanch replied:
@reyang Tagging you just in case you didn't see this. Some food for thought as we improve/refactor the build-up. The ordering problem is similar to [ASP.NET Core Middleware](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-3.1#middleware-order). Not much provided there, other than guidance that order is critical.

Interesting ideas about distinguishing "exporting processors" versus "enriching processors" versus other types.

reyang replied:
Reminded me of [TSR](https://en.wikipedia.org/wiki/Terminate_and_stay_resident_program) and [IRP](https://www.microsoftpressstore.com/articles/article.aspx?p=2201309&seqNum=3).

alanwest commented on docs/trace/enriching-activity-processor/README.md:
````diff
@@ -0,0 +1,118 @@
+Enrichment scopes may be nested.
+
+```csharp
+using EnrichmentScope.Begin(
+    target: EnrichmentScopeTarget.AllChildren,
+    enrichmentAction: a => a.AddTag("mycompany.customer_id", 5678))
+{
````
Cool, I like what is demonstrated here!

alanwest commented on src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs:
```diff
@@ -90,6 +90,7 @@ public override void OnStartActivity(Activity activity, object payload)
 
             activity.SetKind(ActivityKind.Client);
             activity.DisplayName = HttpTagHelper.GetOperationNameForHttpMethod(request.Method);
+            activity.SetCustomProperty("HttpHandler.Request", request);
```
How does this relate to the enrichment stuff?

CodeBlanch replied:
In the demo I wanted to enrich by some context stuff and by adding userAgent from request & contentType from response. Surprised to notice that the raw objects were not available (they are for .NET Framework). So this is making the raw objects available to anyone downstream that might be interested.

CodeBlanch commented:
Had a chat with @cijothomas. I'm going to take a stab at supporting this via CorrelationContext/Baggage.

CodeBlanch commented:
If anyone lands here looking for a way to make this work I have an updated version (for 1.0) here: https://www.nuget.org/packages/Macross.OpenTelemetry.Extensions
